java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -170,21 +207,78 @@ class DatabaseImpl(
     }
 
     /**
-     * Returns a map of field name to ID for each field in the given schema [TypeId].
+     * Returns a map of field name to field ID and type ID, for each field in the given schema
+     * [TypeId].
      *
      * Call [getSchemaTypeId] first to get the [TypeId].
      */
     @VisibleForTesting
-    fun getSchemaFieldIds(schemaTypeId: TypeId): Map<FieldName, FieldId> {
+    fun getSchemaFields(schemaTypeId: TypeId): Map<FieldName, SchemaField> {
         // TODO: Use an LRU cache.
-        val fieldIds = mutableMapOf<FieldName, FieldId>()
+        val fields = mutableMapOf<FieldName, SchemaField>()
         readableDatabase.rawQuery(
-            "SELECT name, id FROM fields WHERE parent_type_id = ?",
+            "SELECT name, id, type_id FROM fields WHERE parent_type_id = ?",
             arrayOf(schemaTypeId.toString())
         ).forEach {
-            fieldIds[it.getString(0)] = it.getLong(1)
+            fields[it.getString(0)] = SchemaField(
+                fieldName = it.getString(0),
+                fieldId = it.getLong(1),
+                typeId = it.getLong(2)
+            )
+        }
+        return fields
+    }
+
+    /**
+     * Returns the ID of the given value from the appropriate primitive table.
+     *
+     * Booleans don't have a primitive table, they will just be returned as either 0 or 1.
+     */
+    @VisibleForTesting
+    fun getPrimitiveValueId(value: Any?, fieldId: FieldId): FieldValueId {
```

Comment:
Could be good to add a comment suggesting we determine how to cache the values.

Reply:
Done